lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeDynamicTypeResolution.cpp

```diff
@@ -699,27 +700,31 @@ SwiftLanguageRuntimeImpl::emplaceClangTypeInfo(
     CompilerType clang_type, llvm::Optional<uint64_t> byte_size,
     llvm::Optional<size_t> bit_align,
     llvm::ArrayRef<swift::reflection::FieldInfo> fields) {
-  std::lock_guard<std::recursive_mutex> locker(m_clang_type_info_mutex);
+  const std::lock_guard<std::recursive_mutex> locker(m_clang_type_info_mutex);
   if (!byte_size || !bit_align) {
     m_clang_type_info.insert({clang_type.GetOpaqueQualType(), llvm::None});
     return nullptr;
   }
   assert(*bit_align % 8 == 0 && "Bit alignment no a multiple of 8!");
   auto byte_align = *bit_align / 8;
   // The stride is the size rounded up to alignment.
-  size_t byte_stride = llvm::alignTo(*byte_size, byte_align);
+  const size_t byte_stride = llvm::alignTo(*byte_size, byte_align);
+  auto extra_inhabitants = 
+      clang_type.IsPointerOrReferenceType()
+          ? swift::swift_getHeapObjectExtraInhabitantCount()
```

Comment:
Are all imported pointers types, heap objects?

Reply:
Maybe we need to restrict this to something like ObjC object pointers. Augusto — you could create a quick counterexample like this:

header.h
void *getPointer() { return (void*)-1ULL; }

test.swift
let ptr = getPointer()
let optional : <type of ptr>? = ptr

in this case `optional` is an imported pointer that definitely uses all bits and where it wouldn't be safe to store the optional enum discriminator inside it.

Reply:
C pointers get imported as one of the `Unsafe*Pointer` types, which exclude `NULL`, so the optional `nil` can use `NULL` as its storage.

However, I'm not sure how such things show up at this level. Those types are actually Swift structs, with their pointer nature being down to the implementation of those structs. Definitely something to double check.

Reply:
I'm actually a bit confused now. From Clang's POV `CGImage` is a typedef to `CGImageRef`, and `CGImageRef` seems to be a C struct. The compiler knows somehow that there are extra inhabitants, but I'm unsure how it gets there so I can follow the same logic here.

Reply:
`CGImageRef` is a typedef to `CGImage*`. Our documentation seems to be missing the star, which is unfortunate... trust the header instead, I guess. The typedef also has the `objc_bridge` attribute on it, which tells Swift that it can be treated like an object pointer.

Reply:
I _believe_ the only heap allocated types that should be going through the external provider are types special cased in the compiler, so I added an `lldb_assert` to make sure this assumption is correct.